# Importing Libraries

In [2]:
# Import libraries
import pandas as pd
import numpy as np
import os

# Importing Data Bases

In [4]:
# Import orders.csv
# Create a string of the path
# Define the path variable
path=r'/Users/douniaelyoussoufi/Achievement 4 '
# Load the Orders CSV file directly using os.path.join
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'orders.csv'), index_col=False)

In [5]:
# Import products.csv
# Load the Products CSV file directly using os.path.join
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

# Step 1

# Data Wrangling Procedures

In [8]:
# Dropping 'eval_set' column from orders.csv
df_ords = df_ords.drop(columns = ['eval_set'])

In [10]:
# Example of finding the Missing (NaN) Values
df_ords['days_since_prior_order'].value_counts(dropna = False)

days_since_prior_order
30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: count, dtype: int64

In [12]:
# Renaming the 'order_dow' column from orders.csv
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [14]:
# Checking the changes made
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [16]:
# Changing the data type of the 'order_id' from orders.csv
df_ords['order_id'] = df_ords['order_id'].astype('str')
# Checking the data type of 'order_id from orders.csv
df_ords['order_id'].dtype

dtype('O')

In [18]:
# Import Departments.csv data set
# Load the Departments CSV file directly using os.path.join
df_dep = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'departments.csv'), index_col=False)
# Previewing the first rows of the dataframe
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [20]:
# Transposing the DataFrame of the Departments.csv with the T Functio
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [22]:
# Creating a new dataframe from the transposed version
df_dep_t = df_dep.T
df_dep_t

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [24]:
# Resetting the index of the dataframe
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [26]:
# Creating a new header for the dataframe
new_header = df_dep_t.iloc[0]
new_header

0    department
Name: department_id, dtype: object

In [28]:
# Creating a new dataframe that only copies over rows beyond the first row
df_dep_t_new = df_dep_t[1:]
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [30]:
# Set the new header row as the df_dep header
df_dep_t_new.columns = new_header
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

# Data Dictionary

In [33]:
# Creating a data dictionary
data_dict = df_dep_t_new.to_dict('index')
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [35]:
# Preview of the first few records from the products dataset
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [37]:
# Retrieving department information for department ID 19
print(data_dict.get('19'))

{'department': 'snacks'}


# Subsetting

In [40]:
# Subset for snacks department
df_snacks =  df_prods[df_prods['department_id']==19]
df_snacks.head()

product_id                                 product_name  aisle_id  \
0            1                   Chocolate Sandwich Cookies        61   
15          16                Mint Chocolate Flavored Syrup       103   
24          25      Salted Caramel Lean Protein & Fiber Bar         3   
31          32                Nacho Cheese White Bean Chips       107   
40          41  Organic Sourdough Einkorn Crackers Rosemary        78   

    department_id  prices  
0              19     5.8  
15             19     5.2  
24             19     1.9  
31             19     4.9  
40             19     6.5

# Task Analysis Procedures

# Step 2

Another identifier variable in the df_ords dataframe that doesn’t need to be included in numeric analysis is likely user_id. This column uniquely identifies users and is not typically analyzed as a numeric variable (e.g., you wouldn't compute sums or averages of user_id values). Instead, it is better represented as a string.

In [45]:
# Convert user_id to string type
df_ords['user_id'] = df_ords['user_id'].astype(str)
# Verify the change
df_ords.dtypes

order_id                   object
user_id                    object
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

# Step 3

Another variable with an unintuitive name in the df_ords dataframe could be days_since_prior_order. While descriptive, it might be lengthy and could be renamed to something more concise like days_prior_order.

In [49]:
# Rename the variable without overwriting the original dataframe
df_ords_renamed = df_ords.rename(columns={'days_since_prior_order': 'days_prior_order'})
# Display the updated column names
df_ords_renamed.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_prior_order'],
      dtype='object')

# Step 4

To determine the busiest hour for placing orders, we need to analyze the frequency of the order_hour_of_day column.

In [53]:
# Calculate the frequency of each hour
hourly_order_counts = df_ords['order_hour_of_day'].value_counts().sort_index()
# Find the busiest hour
busiest_hour = hourly_order_counts.idxmax()
busiest_hour_count = hourly_order_counts.max()
# Display the findings
busiest_hour_info = f"The busiest hour for placing orders is {busiest_hour} with {busiest_hour_count} orders."
busiest_hour_info

'The busiest hour for placing orders is 10 with 288418 orders.'

# Step 5

To determine the meaning behind a value of 4 in the department_id column within the df_prods dataframe, we would need to reference the data dictionary provided for the dataset.

In [57]:
# The data dictionary 
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [59]:
print(data_dict.get('4'))

{'department': 'produce'}


# Step 6

Creating a subset of the dataframe containing only the required information about breakfast item sales

In [63]:
# Filter the dataframe to only include breakfast items, note that breakfast is the departement 14 in data dictionary
breakfast_items = df_prods[df_prods['department_id']==14]
# View the subset
breakfast_items.head()

product_id                                      product_name  aisle_id  \
27           28                                 Wheat Chex Cereal       121   
33           34                                               NaN       121   
67           68                           Pancake Mix, Buttermilk       130   
89           90                                      Smorz Cereal       121   
210         211  Gluten Free Organic Cereal Coconut Maple Vanilla       130   

     department_id  prices  
27              14    10.1  
33              14    12.2  
67              14    13.7  
89              14     3.9  
210             14     3.6

# Step 7

Creating a subset of the dataframe containing the required information about alcohol, deli, beverages, and meat/seafood departements.

In [67]:
# Filter the dataframe to include only the relevant departments
dinner_party_items = df_prods.loc[df_prods['department_id'].isin([5, 7, 12, 20])] # 5 is alcohol, 7 is beverages, 12 is meat/seafood, 20 is deli
# View the subset
dinner_party_items

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

# Step 8

The last dataframe has 7650 rows

# Step 9

Extract all the information about the customer with the user_id of "1" from the df_ords dataframe.

In [73]:
# Filter the df_ords dataframe to get all information for the customer with user_id = 1
user_1_data = df_ords.loc[df_ords['user_id'] == '1']
# View the resulting subset
user_1_data

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1   2398795       1             2                   3                  7   
2    473747       1             3                   3                 12   
3   2254736       1             4                   4                  7   
4    431534       1             5                   4                 15   
5   3367565       1             6                   2                  7   
6    550135       1             7                   1                  9   
7   3108588       1             8                   1                 14   
8   2295261       1             9                   1                 16   
9   2550362       1            10                   4                  8   
10  1187899       1            11                   4                  8   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0

# Step 10

To analyze the behavior of the customer with user_id = 1, we can provide some basic statistics based on the data available in df_ords. These stats will give us insight into the ordering habits, such as frequency of orders or the timing of orders.

In [76]:
# Total number of orders placed by user_id = 1
total_orders = user_1_data['order_id'].count()
print(f"Total number of orders: {total_orders}")

Total number of orders: 11


In [237]:
# Calculate the average number of days between orders
avg_days_between_orders = user_1_data['days_since_prior_order'].mean()
print(f"Average days between orders: {avg_days_between_orders} days")

Average days between orders: 19.0 days


# Exporting Dataframes

In [79]:
df_ords

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0        2539329       1             1                   2                  8   
1        2398795       1             2                   3                  7   
2         473747       1             3                   3                 12   
3        2254736       1             4                   4                  7   
4         431534       1             5                   4                 15   
...          ...     ...           ...                 ...                ...   
3421078  2266710  206209            10                   5                 18   
3421079  1854736  206209            11                   4                 10   
3421080   626363  206209            12                   1                 12   
3421081  2977660  206209            13                   1                 12   
3421082   272231  206209            14                   6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [241]:
# Export the df_ords dataframe as “orders_wrangled.csv” 
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_wrangled.csv'))

In [243]:
# Export the df_dep_t_new dataframe as “departments_wrangled.csv” 
df_dep_t_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'departments_wrangled.csv'))